### PR TITLE
[TEP-0135] Revert PVC creation

### PR DESCRIPTION
**Pull request by QuanZhang-William**
Part of [#6740] and unblocks [#6635]. `PVCs` are created if a user specifies `VolumeClaimTemplate` as the source of a `pipelinerun workspace`. Prior to this change, such `pvcs` are created via `affinity assistant statefulset` when `affinity assistant` is enabled (in both `workspaces` or `pipelineruns` coschedule mode).

To delete such pvcs when the owning `pipelinerun` is completed, we must explicitly delete those pvcs in the reconciler since the owner of such pvcs is the `affinity assistant statefulset` instead of the `pipelinerun`. The problem of such deletion mechanism is that such `pvcs` are left in `terminating` state when the owning `pipelinerun` is `completed` but not `deleted`. This is because the pvcs are protected by `kubernetes.io/pvc-protection` `finalizer`, which does not allow the `pvcs` to be deleted until the `pipelinerun` consuming the `pvc` is deleted. Leaving pvcs in `terminating` state is confusing to cluster operators. Instead of changing the pvc deletion behavior in such backward incompatible way, it is better to make it configurable so that it is backward compatible, as prototyped in [#6635].

This commit reverts the pvc creation behavior `per-workspace` coschedule mode, which changes the owner of the `pvcs` back to the `pipelinerun` instead of the `affinity assistant statefulset`. After the commit, the pvcs created by specifying `VolumeClaimTemplate` are left in `bounded` state instead of `terminating`. This commit is the prerequisite of [#6635].

This commit does NOT reverts the pvc creation behavior `per-pipelinerun` coschedule mode as we will enforce the deletion of pvcs when the owning `pipelinerun` is completed. This is a better practice and there is no backward compatability concern since `per-pipelinerun` coschedule mode is a new feature.

[#6740]: https://github.com/tektoncd/pipeline/issues/6740
[#6635]: https://github.com/tektoncd/pipeline/pull/6635

# Changes

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [ ] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
TEP-0135: Revert the owner of `PVCs` created by `pipelinerun VolumeClaimTemplate` back to `pipelinerun`. The `PVCs` bounded to the `pipelinerun` is now in `bounded` state when the `pipelinerun` is completed but not deleted.
```

### CODE REVIEW

**Comment by QuanZhang-William** [comment]
cc @lbernick @pritidesai @afrittoli

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant.go:189]: would it make sense to revert only this change?

Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:189]: Revert this part only will resulting in pvcs not deleted when the pipelinerun is deleted though

Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:189]: And I'm not sure https://github.com/tektoncd/pipeline/pull/6635 can work if only reverting this part 🤔

Comment (lbernick) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:189]: We could add a finalizer for PipelineRuns that deletes any affinity assistant PVCs when the PipelineRun is deleted. Also, can you elaborate on what #6635 might not work?

Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:189]: Yeah. If we just revert this part, the pvcs won't be deleted after this PR is merged. But if we can merge #6635 in the same release to delete the `pvcs` when the `pr` is deleted, it should be fine probably.

Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:189]: /hold

Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:189]: I think we still need to call APIs to delete `pvcs` at `pr` completion/deletion time anyways. 

The advantage of this approach is that there is no extra work to delete the `pvcs` at `pr` deletion time, as the owner of the `pvcs` are the `prs` not the `aa statefulset` (i.e. k8s automatically deletes the pvcs when the owning prs are deleted). 

In addition, this patch fully reverts the pvcs behaviours in 1 pull request. If we just revert the highlighted part but not merge the followup prs in one release, users will still see the different behavior (i.e. `pvcs` are not deleted when `pr` is deleted). 

I personally feel this full revert could be a better solution. I'm open to the partial solution if we can make sure to merge all the followup prs in one pr, but I don't see much advantage.

WDYT @lbernick @pritidesai @afrittoli

Comment (lbernick) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:189]: OK, I think it's a fair point that finalizers don't have much advantage over owner references, and have more unknowns since we haven't tried that approach before. this PR SGTM.

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:91]: This is a good point. I think `affinityAssistantStatefulSet ` should have the sole responsibility to create an `StatefulSet` based on the given input (so usually we don't change it), and all our business logic should go to `createOrUpdateAffinityAssistantPerAABehavior` and `createOrUpdateAffinityAssistant`. 

Let me try to find some better variable name in `createOrUpdateAffinityAssistantsPerAABehavior` and put some comments

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:189]: Thanks, Lee! 
/hold cancel

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/pipelinerun.go]: SGTM, updated

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant_test.go]: SGTM, updated

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/pipelinerun.go:618]: At this point, we are creating PVCs from VolumeClaimTemplate before creating AA if it the behavior is `coschedule workspace` or `disabled`. We skip such logic for `coschedule pipelinerun` mode

Comment (lbernick) [reply pkg/reconciler/pipelinerun/pipelinerun.go:618]: hm I think I understand; possible suggestion to make the code a bit cleaner: what if we move PVC creation to `createOrUpdateAffinityAssistants`? It would help that function make fewer assumptions about what pvcs exist and make handling the disabled case less redundant. Up to you.

Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/pipelinerun.go:618]: hmmm, as a developer reading code, I wouldn't expect the pvc creation is in the `createOrUpdateAffinityAssistants` especially when AA is disabled. I feel it could be better to separate concern bettween PVC creation and AA?

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:95]: I think the variable names are accurate, so I added some doc for this part, PTAL 😁 !

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/pipelinerun.go:618]: Synced with @lbernick offline, we agreed to put the PVC creation logic in the Affinity Assistant creation function to make it atomic.

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant.go:56]: nit: `perAABehavior` is a bit redundant with the function arguments; we could shorten the function name to createOrUpdateAffinityAssistantsAndPVCs or `handleVolumeScheduling`

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant.go:84]: CreatePVCsForWorkspacesWithoutAffinityAssistant will fail if it tries to create PVCs that already exist, but createOrUpdateAffinityAssistantsAndPVCsPerAABehavior runs on every reconcile loop. Will this work past the first reconcile loop?

Comment (lbernick) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:84]: Also, we need to update the docstring of CreatePVCsForWorkspacesWithoutAffinityAssistant https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/volumeclaim/pvchandler.go#L54-L59

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (lbernick) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:102]: Here's how I would refactor this (involves rewriting CreatePVCsForWorkspacesWithoutAffinityAssistant):

```
case aa.AffinityAssistantPerWorkspace:
                 ...
		for claimTemplate, workspaceBinding := range claimTemplatesToWorkspace {
			aaName := getAffinityAssistantName(workspaceBinding, pr.Name)
                         pvc, err := c.pvcHandler.GetOrCreatePVC(workspaceBinding, pr)
                         < handle err>
                         err := c.createOrUpdateAffinityAssistant(ctx, aaName, pr, nil, []corev1.PersistentVolumeClaimVolumeSource{pvc}, unschedulableNodes)
                         < handle err>
```

However, I know this PR is just meant to revert deletion behavior for the existing affinity assistant, and I want to keep this PR scoped to the minimal set of changes to do that. (I recognize I've already asked for some refactoring, but this chunk of code can be hard to grapple with :)) Maybe refactoring changes could be split into a separate PR?

**Review by lbernick — COMMENTED** [review]
thanks @QuanZhang-William, just two additional comments:
1. could you please not re-run tests without some explanation of why you think they are flaky? Linking to an issue tracking the flake is ideal.
2. The reason we didn't catch this behavior change originally is due to a lack of integration tests, and I think the way k8s handles deletion of PVCs when the PV still exists would be hard to mock in a unit test without being very brittle. I think we should add integration tests that ensure PVCs are in the desired state after completion of the pipelinerun.

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:84]: No it will be only executed once, as it is guarded by `pipelineRunFacts.State.IsBeforeFirstTaskRun()`. The `CreatePVCsForWorkspacesWithoutAffinityAssistant ` also checks the existence of the PVCs

**Comment by QuanZhang-William** [comment]
> thanks @QuanZhang-William, just two additional comments:
> 
> 1. could you please not re-run tests without some explanation of why you think they are flaky? Linking to an issue tracking the flake is ideal.

SGTM!
> 2. The reason we didn't catch this behavior change originally is due to a lack of integration tests, and I think the way k8s handles deletion of PVCs when the PV still exists would be hard to mock in a unit test without being very brittle. I think we should add integration tests that ensure PVCs are in the desired state after completion of the pipelinerun.

I think we add the integration test when we work on changing the PVC deletion behavior PR since the deletion behavior is pending change and we need to modify the integration test at time anyways.

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go]: Renamed to createOrUpdateAffinityAssistantsAndPVCs

**Comment by QuanZhang-William** [comment]
> > > 2. The reason we didn't catch this behavior change originally is due to a lack of integration tests, and I think the way k8s handles deletion of PVCs when the PV still exists would be hard to mock in a unit test without being very brittle. I think we should add integration tests that ensure PVCs are in the desired state after completion of the pipelinerun.
> > 
> > 
> > I think we add the integration test when we work on changing the PVC deletion behavior PR since the deletion behavior is pending change and we need to modify the integration test at time anyways.
> 
> I disagree-- the purpose of this PR is to change the deletion behavior, and I feel I'd have to test it out locally to ensure the PVCs are still in "bound" state on completion of the pipelinerun. This signals to me we need an integration test for this behavior. When the deletion behavior is modified, the tests can be easily modified as well; this will make it even easier to understand the impacts of the change.

SGTM, I have added an integration test to test the AA behavior from end to end. PTAL

**Comment by QuanZhang-William** [comment]
/test pull-tekton-pipeline-integration-tests 
`status code 500 Internal Server Error` when retrieving image

File: test/affinity_assistant_test.go
Comment (lbernick) [inline test/affinity_assistant_test.go:34]: This looks great! Thanks Quan.

File: pkg/reconciler/volumeclaim/pvchandler.go
Comment (lbernick) [inline pkg/reconciler/volumeclaim/pvchandler.go:60]: I'm not a big fan of docstrings that explain how a function is called. I think it would be sufficient to explain that it creates PVCs for workspaces with volumeClaimTemplates

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant_test.go:289]: It looks like this is just checking the PVC exists? is it expected that the return value from the client is ignored?

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (lbernick) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:102]: I'm approving this PR but I have a strong preference that refactoring changes happen first. typically refactoring changes that get deferred to later just get deprioritized.

**Review by lbernick — APPROVED** [review]

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:102]: Hi Lee, I did some investigation and realized that doing such refactoring in this PR will introduce a bunch of code changes. This makes this PR quite big and harder to review.

To address your concern, I have created an issue tracking it and assigned to myself: https://github.com/tektoncd/pipeline/issues/6915 (will make sure to prioritize). Merging this PR as it is can help unblocking the followup PRs where I will add e2e support for `coschedule pipelineruns` mode, and I can work on the refactoring work in parallel (and the reviews for the followup PRs can happen in parallel). 

WDTY?

Comment (lbernick) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:102]: Thank you! I appreciate it!

File: pkg/reconciler/volumeclaim/pvchandler.go
Comment (QuanZhang-William) [reply pkg/reconciler/volumeclaim/pvchandler.go]: I have removed the usage docstring for now. Will further look into the explanation in the followup refactor PR

File: test/affinity_assistant_test.go
Comment (QuanZhang-William) [reply test/affinity_assistant_test.go:36]: SGTM, docstring added

**Review by Yongxuanzhang — APPROVED** [review]

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant_test.go:188]: Good catch, fixed!

**Review by JeromeJu — APPROVED** [review]
Thanks for answering the concerns on testing offline and adding the excellent commit message. Discussed couple of implementation choices with @QuanZhang-William offline.

/lgtm

**Pull request merged**
